src/Features/Core/SolutionCrawler/WorkCoordinator.AsyncWorkItemQueue.cs

```diff
@@ -99,6 +99,15 @@ public bool AddOrReplace(WorkItem item)
                     }
                 }
 
+                public void RequestCancellationOnRunningTasks()
+                {
+                    lock (_gate)
+                    {
+                        // request to cancel all running works
+                        _cancellationMap.Do(p => p.Value.Cancel());
```

Comment:
Shouldn't we be invoking Cancel_NoLock here to ensure the entry is also deleted from the _cancellationMap? Otherwise, the existing AddOrReplace(WorkItem) requests waiting on the gate will end up invoking Cancel_NoLock as soon as they can lock gate and invoke Cancel_NoLock, which will try to Cancel the source again.